syntax = "proto3";
package bastionlab_polars;

message ReferenceRequest {
    string identifier = 1;
}

message ReferenceResponse {
    string identifier = 1;
    string header = 2;
}

message ReferenceList {
    repeated ReferenceResponse list = 1;
}

message SendChunk {
    bytes data = 1;
    string policy = 2;
    string metadata = 3;
}

message FetchChunk {
    oneof body {
        bytes data = 1;
        string pending = 2;
        string warning = 3;
    }
}

message Query {
    string composite_plan = 1;
}

message Empty {}

message SplitRequest {
    repeated ReferenceRequest arrays = 1;
    float train_size = 2;
    float test_size = 3;
    bool shuffle = 5;
    optional uint64 random_state = 6;
}

service PolarsService {
    rpc SendDataFrame (stream SendChunk) returns (ReferenceResponse) {}
    rpc RunQuery (Query) returns (ReferenceResponse) {}
    rpc FetchDataFrame (ReferenceRequest) returns (stream FetchChunk) {}
    rpc ListDataFrames (Empty) returns (ReferenceList) {}
    rpc GetDataFrameHeader (ReferenceRequest) returns (ReferenceResponse) {}
    rpc PersistDataFrame (ReferenceRequest) returns (Empty) {}
    rpc DeleteDataFrame (ReferenceRequest) returns (Empty) {}
    rpc Split(SplitRequest) returns (ReferenceList) {}
}
